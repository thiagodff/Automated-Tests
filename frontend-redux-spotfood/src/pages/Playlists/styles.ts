import styled from 'styled-components';

export const Container = styled.div`
  height: 100%;
  margin: 0 5%;
`;

export const CenteredContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 80%;
`;

export const Filters = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;

  > div,
  > span {
    width: 17rem;
    height: 3.2rem;
  }

  > * {
    margin: 2rem 1rem;

    input {
      width: 100%;
    }
  }
`;

export const PlaylistContainer = styled.div`
  margin-top: 3rem;
  h2 {
    text-align: center;
    font-weight: bold;
  }
`;

export const PlaylistList = styled.ul`
  list-style-type: none;
  display: flex;
  justify-content: center;
  flex-wrap: wrap;
  align-items: flex-start;
`;

export const Playlist = styled.li`
  margin: 0 1rem 1rem 1rem;
`;
