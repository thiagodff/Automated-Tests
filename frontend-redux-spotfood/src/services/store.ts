import { store } from '../store';

export function getSpotifyAccessToken() {
  return store.getState().auth.access_token;
}

export function getSpotifyAccessTokenType() {
  return store.getState().auth.token_type;
}