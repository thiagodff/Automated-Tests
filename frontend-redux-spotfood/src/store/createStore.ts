import { createStore, applyMiddleware, Middleware, Reducer } from 'redux';
import { AuthAction, AuthState } from './modules/auth/types';
import { PlaylistsAction, PlaylistsState } from './modules/playlists/types';
import { PersistState } from 'redux-persist';

export interface StoreState {
  _persist: PersistState;
  auth: AuthState;
  playlists: PlaylistsState;
}

export type StoreAction = AuthAction & PlaylistsAction;

export default (
  reducers: Reducer<StoreState, StoreAction>,
  middlewares: Middleware[]
) => {
  const enhancer = applyMiddleware(...middlewares);

  return createStore(reducers, enhancer);
};
