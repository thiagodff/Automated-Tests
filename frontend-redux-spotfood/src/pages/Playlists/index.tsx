import React, { useCallback, useEffect, useMemo, useState } from 'react';

import debounce from 'lodash/debounce';
import { useSelector, useDispatch } from 'react-redux';
import { Select, DatePicker, Input, Card, Typography, Spin, InputNumber } from 'antd';

import { getPlaylistsRequest } from '../../store/modules/playlists/actions';
import { StoreState } from '../../store/createStore';
import { FiltersState, Playlist as PlaylistType } from '../../store/modules/playlists/types';

import { Container, PlaylistList, Playlist, Filters, PlaylistContainer, CenteredContainer } from './styles';

interface DynamicFilterState {
  [key: string]: string | number | undefined, offset: number; limit: number
}

const initialFilterState: DynamicFilterState = {
  locale: undefined,
  country: undefined,
  timestamp: undefined,
  limit: 20,
  offset: 0,
}

const { Meta } = Card;
const { Title } = Typography;

const TIME_TO_REFRESH_PAGE_IN_MILLISECONDS = 30000;

const Playlists = () => {
  const dispatch = useDispatch();

  const [filtersState, setFiltersState] = useState<FiltersState>(initialFilterState);
  const [nameFilter, setNameFilter] = useState<string>();

  const handleSetNameFilter = useCallback(
    debounce((value) => setNameFilter(value), 250),
    []);

  const { playlists, filters, fetchPlaylistsStatus } = useSelector((state: StoreState) => state.playlists);

  const filteredPlaylists: PlaylistType[] = useMemo(() => {
    if (!playlists?.playlists) {
      return [];
    }

    if (nameFilter) {
      const filterNameRegex = new RegExp(nameFilter, 'i');
      return playlists.playlists.items.filter((playlist) => playlist.name.match(filterNameRegex));
    }

    return playlists?.playlists.items;
  }, [nameFilter, playlists]);

  const getPlaylists = useCallback(() => {
    dispatch(getPlaylistsRequest({ filters: filtersState }));
  }, [dispatch, filtersState]);

  const handleFilter = useCallback(
    debounce(({ key, value }: { key: string; value?: string }) => setFiltersState((prevState: FiltersState) => ({ ...prevState, [key]: value || initialFilterState[key] })), 250
    ), []);

  useEffect(() => {
    getPlaylists();
  }, [getPlaylists, dispatch]);

  useEffect(() => {
    const intervalId = setInterval(() => {
      window.location.reload();
    }, TIME_TO_REFRESH_PAGE_IN_MILLISECONDS)

    return () => clearInterval(intervalId);
  }, [])

  const getPlaylistCards = useCallback(() => (
    <PlaylistContainer>
      <Title level={2}>{playlists?.message}</Title>
      <PlaylistList>
        {
          filteredPlaylists.map((playlist) => (
            <Playlist
              key={playlist.id}
            >
              <Card
                hoverable
                style={{ width: '25rem' }}
                cover={<img src={playlist.images[0].url} alt="Playlist" />}
              >
                <Meta title={playlist.name} description={playlist.description} />
              </Card>
            </Playlist>
          ))
        }
      </PlaylistList>
    </PlaylistContainer>
  ), [filteredPlaylists, playlists])

  const getContent = useCallback(() => {
    if (fetchPlaylistsStatus === 'fetching' || fetchPlaylistsStatus === 'notFetched') {
      return (
        <CenteredContainer>
          <Spin size="large" />
        </CenteredContainer>
      );
    }

    if (fetchPlaylistsStatus === 'error') {
      return (
        <CenteredContainer>
          <Title level={3}>Houve um erro ao tentar buscar novas playlists.</Title>
        </CenteredContainer>
      );
    }

    if (filteredPlaylists.length === 0) {
      return (
        <CenteredContainer>
          <Title level={3}>Nenhum resultado encontrado. Tente mudar os filtros de busca.</Title>
        </CenteredContainer>
      );
    }

    return getPlaylistCards();
  }, [fetchPlaylistsStatus, filteredPlaylists.length, getPlaylistCards])

  return (
    <Container>
      <Filters>
        <Title level={4}>Filtros:</Title>
        {filters?.selectFilters.map((filter) => (
          <Select
            key={filter.id}
            placeholder={`Selecione o(a) ${filter.name}`}
            options={filter.values} size="middle"
            allowClear
            onSelect={(value) => handleFilter({ key: filter.id, value: String(value) })}
          />
        ))}
        {filters?.typedFilters.map((filter) => (
          filter.validation.entityType === 'DATE_TIME' ? (
            <DatePicker
              key={filter.id}
              placeholder={`Selecione o(a) ${filter.name}`}
              size="middle"
              allowClear
              format={filter.validation.pattern}
              onChange={(event) => handleFilter({ key: filter.id, value: event?.format() })}
            />
          ) : (
              <InputNumber
                key={filter.id}
                placeholder={`Selecione o(a) ${filter.name}`}
                size="middle"
                onChange={(value) => handleFilter({ key: filter.id, value: String(value) })}
              />
            )))}
        <Input
          placeholder="Busque por uma playlist"
          size="middle"
          allowClear
          onChange={(event) => handleSetNameFilter(event.target.value)}
        />
      </Filters>

      {getContent()}
    </Container>
  );
}

export default Playlists;
