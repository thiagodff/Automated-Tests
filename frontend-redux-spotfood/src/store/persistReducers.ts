import storage from 'redux-persist/lib/storage';
import { persistReducer } from 'redux-persist';
import { Reducer } from 'redux';
import { StoreAction, StoreState } from './createStore';

export default (reducers: Reducer<StoreState, StoreAction>) => {
  const persistedReducer = persistReducer(
    {
      key: 'spotifood',
      storage,
      whitelist: ['auth', 'playlists'],
    },
    reducers
  );

  return persistedReducer;
};
