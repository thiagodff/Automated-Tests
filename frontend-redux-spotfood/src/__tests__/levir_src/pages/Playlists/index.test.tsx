import React from 'react';
import { render } from '@testing-library/react';
import Playlists from '../../../../pages/Playlists';
import { Provider } from 'react-redux';
import { store } from '../../../../store'

describe('[PAGE]: Playlists', () => {
  it('It should be render Playlists page', () => {
    const { container } = render(
      <Provider store={store}>
        <Playlists />
      </Provider>);
    expect(container).toMatchSnapshot();
  });
});
