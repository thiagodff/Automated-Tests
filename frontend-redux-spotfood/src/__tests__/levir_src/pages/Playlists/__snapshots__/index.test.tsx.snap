// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[PAGE]: Playlists It should be render Playlists page 1`] = `
<div>
  <div
    class="sc-bdnylx fGgZzb"
  >
    <div
      class="sc-dlnjPT ghBSrS"
    >
      <h4
        class="ant-typography"
      >
        Filtros:
      </h4>
      <span
        class="ant-input-affix-wrapper"
      >
        <input
          class="ant-input"
          placeholder="Busque por uma playlist"
          type="text"
          value=""
        />
        <span
          class="ant-input-suffix"
        >
          <span
            aria-label="close-circle"
            class="anticon anticon-close-circle ant-input-clear-icon-hidden ant-input-clear-icon"
            role="button"
            tabindex="-1"
          >
            <svg
              aria-hidden="true"
              class=""
              data-icon="close-circle"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
              />
            </svg>
          </span>
        </span>
      </span>
    </div>
    <div
      class="sc-gtssRu flxmAQ"
    >
      <div
        class="ant-spin ant-spin-lg ant-spin-spinning"
      >
        <span
          class="ant-spin-dot ant-spin-dot-spin"
        >
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
          <i
            class="ant-spin-dot-item"
          />
        </span>
      </div>
    </div>
  </div>
</div>
`;
